File: demo-app/lib/models/friend/friend.dart
class Friend {
  final String ownerUserId;
  final String userId;
  final String username;
  final String? email;
  final String? nakamaUserId; // Nakama user ID for chat
  final String? channelId; // Nakama DM channel ID (format: 4.{userId1}.{userId2})
  final String friendshipId;
  final int createdAt;

  Friend({
    required this.ownerUserId,
    required this.userId,
    required this.username,
    this.email,
    this.nakamaUserId,
    this.channelId,
    required this.friendshipId,
    required this.createdAt,
  });

  factory Friend.fromJson(Map<String, dynamic> json) {
    return Friend(
      ownerUserId: json['owner_user_id'] as String? ?? '',
      userId: json['user_id'] as String? ?? '',
      username: json['username'] as String? ?? '',
      email: json['email'] as String?,
      nakamaUserId: json['nakama_user_id'] as String?,
      channelId: json['channel_id'] as String?,
      friendshipId: json['friendship_id'] as String? ?? '',
      createdAt: json['created_at'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'owner_user_id': ownerUserId,
      'user_id': userId,
      'username': username,
      'email': email,
      'nakama_user_id': nakamaUserId,
      'channel_id': channelId,
      'friendship_id': friendshipId,
      'created_at': createdAt,
    };
  }

  @override
  String toString() {
    return 'Friend(userId: $userId, username: $username, friendshipId: $friendshipId)';
  }
}

File: key-service/internal/models/friend.go
package models

type FriendRequest struct {
	RequestID    string  `json:"request_id"`
	FromUserID   string  `json:"from_user_id"`
	FromUsername string  `json:"from_username"`
	FromEmail    *string `json:"from_email,omitempty"`
	ToUserID     string  `json:"to_user_id"`
	ToUsername   string  `json:"to_username"`
	Status       string  `json:"status"` // "pending", "accepted", "rejected"
	CreatedAt    int64   `json:"created_at"`
}

type Friend struct {
	OwnerUserID  string  `json:"owner_user_id"` // User who owns this friendship record
	UserID       string  `json:"user_id"`       // The friend's user ID
	Username     string  `json:"username"`
	Email        *string `json:"email,omitempty"`
	NakamaUserID *string `json:"nakama_user_id,omitempty"` // Nakama user ID for chat
	ChannelID    *string `json:"channel_id,omitempty"`     // Nakama DM channel ID (format: 4.{userId1}.{userId2})
	FriendshipID string  `json:"friendship_id"`
	CreatedAt    int64   `json:"created_at"`
}

type SendFriendRequestRequest struct {
	Username *string `json:"username,omitempty"`
	Email    *string `json:"email,omitempty"`
}

type SendFriendRequestResponse struct {
	RequestID string `json:"request_id"`
	Message   string `json:"message"`
}

type AcceptFriendRequestRequest struct {
	RequestID string `json:"request_id" binding:"required"`
}

type AcceptFriendRequestResponse struct {
	FriendshipID string `json:"friendship_id"`
	Message      string `json:"message"`
}

type FriendRequestsResponse struct {
	Requests []FriendRequest `json:"requests"`
}

type FriendsListResponse struct {
	Friends []Friend `json:"friends"`
}

type RejectFriendRequestRequest struct {
	RequestID string `json:"request_id" binding:"required"`
}

type RejectFriendRequestResponse struct {
	Message string `json:"message"`
}

type RemoveFriendRequest struct {
	FriendUserID string `json:"friend_user_id" binding:"required"`
}

type RemoveFriendResponse struct {
	Message string `json:"message"`
}

type SentFriendRequestsResponse struct {
	Requests []FriendRequest `json:"requests"`
}
